dmon.8
.\" Man page generated from reStructuredText.
.
.TH DMON 8 "" "" ""
.SH NAME
dmon \- Daemonize and monitor processes
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBdmon [options] cmd [cmdoptions] [\-\- logcmd [logcmdoptions]]\fP
.SH DESCRIPTION
.sp
The \fBdmon\fP program will launch a program and re\-launch it whenever it
dies. Optionally, the standard output streams of the programs may be piped
into a second program (named \fIlog command\fP), which will receive the output
of the program in its standard input stream. The log command will be also
monitored and re\-launched when it dies.
.SH USAGE
.sp
Command line options:
.INDENT 0.0
.TP
.BI \-C \ PATH\fR,\fB \ \-\-config \ PATH
Read contents from \fIPATH\fP as if they were command line options.
Those will be parsed after the options picked from the
\fBDMON_OPTIONS\fP environment variable and before the options
given in the command line. If given, this option \fBmust\fP be
the first one passed to to \fBdmon\fP\&.
.TP
.BI \-I \ PATH\fR,\fB \ \-\-write\-info \ PATH
Write status changes of monitored processes to \fIPATH\fP, one
status message per line. See the \fI\%status file format\fP section
for details on the format.
.TP
.BI \-p \ PATH\fR,\fB \ \-\-pid\-file \ PATH
Write the PID of the master \fBdmon\fP process to a file in the
specified \fIPATH\fP\&. You can signal the process to interact with
it. (See \fI\%SIGNALS\fP below.)
.TP
.BI \-W \ PATH\fR,\fB \ \-\-work\-dir \ PATH
Change to the directory located at \fIPATH\fP and use it as working
directory. Note that all other paths passed to \fBdmon\fP (except
the configuration file) will be interpreted as relative to the
working directory.
.TP
.BI \-i \ TIME\fR,\fB \ \-\-interval \ TIME
When execution of the process ends with a successful (zero)
exit status, wait for \fITIME\fP seconds before respawning the
process instead of doing it immediately. This can be used to
make \fBdmon\fP behave as el\-cheapo \fIcron(8)\fP replacement. This
option cannot be used along with \fB\-1\fP\&.
.TP
.BI \-t \ TIME\fR,\fB \ \-\-timeout \ TIME
If the process takes longer than \fITIME\fP seconds to complete,
terminate it by sending the \fITERM\fP/\fICONT\fP signal combo. Then
the process will be respawned again. This is useful to ensure
that potentially locking processes which should take less than
some known time limit do not hog the computer. Most likely,
this flag is useful in conjunction with \fB\-1\fP, and with
\fB\-n\fP e.g. when using it in a \fIcron(8)\fP job.
.TP
.BI \-L \ NUMBER\fR,\fB \ \-\-load\-high \ NUMBER
Enable tracking the system\(aqs load average, and suspend the
execution of the command process when the system load goes
over \fINUMBER\fP\&. To pause the process, \fISTOP\fP signal will be
sent to it. You may want to use \fB\-l\fP as well to specify
under which load value the process is resumed, otherwise
when the system load falls below \fINUMBER/2\fP the process will
be resumed.
.TP
.BI \-l \ NUMBER\fR,\fB \ \-\-load\-low \ NUMBER
When using \fB\-L\fP, the command process execution will be
resumed when the system load falls below \fINUMBER\fP, instead of
using the default behavior of resuming the process when the
load falls below half the limit specified with \fB\-L\fP\&.
.TP
.BI \-E \ ENVVAR\fR,\fB \ \-\-environ \ ENVVAR
Manipulates environment variables. Specifying just a variable
name (e.g. \fB\-E foo\fP) as \fIENVVAR\fP will clear it and remove
the variable from the environment. Adding a value will define
the variable (e.g. \fB\-E foo=bar\fP). This option may be
specified multiple times. Environment variables will affect
\fIboth\fP the \fBdmon\fP and the child process; this is intended
behaviour.
.TP
.BI \-u \ UIDGID\fR,\fB \ \-\-cmd\-user \ UIDGID
Executes the command with the credentials of user \fIUID\fP,
and additional group \fIGID\fP specified separated with
semicolons. Both user and group identifiers might be given
as strings or numerically.
.TP
.BI \-U \ UIDGID\fR,\fB \ \-\-log\-user \ UIDGID
Executes the \fBlog\fP command with the credentials of user
\fIUID\fP, and additional group \fIGID\fP specified separated with
semicolons. Both user and group identifiers might be given
as strings or numerically.
.TP
.B \-n\fP,\fB  \-\-no\-daemon
Do not daemonize: \fBdmon\fP will keep working in foreground,
without detaching and without closing its standard input and
output streams. This is useful for debugging and, to a limited
extent, to run interactive programs.
.TP
.B \-1\fP,\fB  \-\-once
Run command only once: if the command exits with a success
status (i.e. exit code is zero), then \fBdmon\fP will exit and
stop the logging process. If the program dies due to a signal
or with a non\-zero exit status, it is respawned. This option
tends to be used in conjunction with \fB\-n\fP, and cannot be
used with \fB\-i\fP\&.
.TP
.BI \-m \ COUNT\fR,\fB \ \-\-max\-respawns \ COUNT
Respawn command m number of times. \fBdmon\fP does not check the
exit status on each respawn. For example, if you want the
command to never respawn, set \fB\-m 0\fP\&. After the set number of
respawns have passed \fBdmon\fP will NOT respawn the cmd.
Instead, \fBdmon\fP will exit and stop the logging process.
.TP
.B \-e\fP,\fB  \-\-stderr\-redir
Redirect both the standard error and standard output streams
to the log command. If not specified, only the standard output
is redirected.
.TP
.B \-s\fP,\fB  \-\-cmd\-sigs
Forward signals \fICONT\fP, \fIALRM\fP, \fIQUIT\fP, \fIUSR1\fP, \fIUSR2\fP and
\fIHUP\fP to the monitored command when \fBdmon\fP receives them.
.TP
.B \-S\fP,\fB  \-\-log\-sigs
Forward signals \fICONT\fP, \fIALRM\fP, \fIQUIT\fP, \fIUSR1\fP, \fIUSR2\fP and
\fIHUP\fP to the log command when \fBdmon\fP receives them.
.TP
.BI \-r \ LIMIT\fR,\fB \ \-\-limit \ LIMIT
Set \fILIMIT\fP for process execution. Limits are specified as
\fBname=value\fP strings, and multiple limits may be set by
using \fB\-r\fP multiple times. The available set of limits
depends on the current operating system, to get a list
\fB\-r help\fP can be used.
.TP
.B \-h\fP,\fB  \-\-help
Show a summary of available options.
.UNINDENT
.sp
Usual log commands include \fIdlog(8)\fP and \fIdslog(8)\fP, which are part of the
\fBdmon\fP suite. Other log commands like \fIrotlog(8)\fP or \fImultilog(8)\fP may be
used as long as they consume data from standard input and do not detach
themsemlves from the controlling process.
.sp
As a convenience, time values passed to \fB\-i\fP, \fB\-t\fP and values of limits
specified with \fB\-r\fP may be given with the following suffixes:
.INDENT 0.0
.IP \(bu 2
\fBm\fP: Minutes, e.g. \fB30m\fP means "30 minutes".
.IP \(bu 2
\fBh\fP: Hours, e.g. \fB4h\fP means "4 hours".
.IP \(bu 2
\fBd\fP: Days, e.g. \fB3d\fP means "3 days".
.IP \(bu 2
\fBw\fP: Weeks, e.g. \fB1w\fP means "1 week".
.UNINDENT
.sp
For size values (bytes) the strings passed to \fB\-r\fP as limits may have the
following suffixes:
.INDENT 0.0
.IP \(bu 2
\fBk\fP: Kilobytes.
.IP \(bu 2
\fBm\fP: Megabytes.
.IP \(bu 2
\fBg\fP: Gigabytes.
.UNINDENT
.SH SIGNALS
.sp
Signals may be used to interact with the monitored processes and \fBdmon\fP
itself.
.sp
The \fBTERM\fP and \fBINT\fP signals are catched by \fBdmon\fP, and they will
make it shut down gracefully: both the main command and the log command
will receive a \fBTERM\fP signal followed by a \fBCONT\fP and they will be
waited for.
.sp
When at least one of \fB\-s\fP or \fB\-S\fP are used, the \fBCONT\fP, \fBALRM\fP,
\fBQUIT\fP, \fBUSR1\fP, \fBUSR2\fP and \fBHUP\fP signals are forwarded to the
managed processes. By default, if none of the options are used, those
signals are ignored.
.SH EXAMPLES
.sp
The following command will supervise a shell which prints a string each
fifth second, and the output is logged to a file with timestamps:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dmon \-n sh \-c \(aqwhile echo "Hello World" ; do sleep 5 ; done\(aq \e
  \-\- dlog logfile
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In order to turn the previous example into a daemon, we only need to
remove the \fB\-n\fP\&. I may be convenient to specify a PID file path:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dmon \-p example.pid \e
  sh \-c \(aqwhile echo "Hello dmon" ; do sleep 5 ; done\(aq \e
  \-\- dlog logfile
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The following example launches the \fIcron(8)\fP daemon with the logging
process running as user and group \fBlog:wheel\fP:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dmon \-p /var/run/crond.pid \-u log:wheel \-e cron \-f
  \-\- dlog /var/log/cron.log
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This example will run a (probably lengthy) backup process, pausing it when
the system load goes above 3.5 and resuming it when the load drops below
1.0:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dmon \-1 \-n \-l 1 \-L 3.5 rsync \-avz ~/ /backup/homedir
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If you have a PID file, terminating the daemon is an easy task:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
kill $(cat example.pid)
.ft P
.fi
.UNINDENT
.UNINDENT
.SH STATUS FILE FORMAT
.sp
When using the \fB\-I\fP \fIPATH\fP option, status updates are written to \fIPATH\fP,
one line per update. The following line formats may be used:
.sp
A process was started by \fBdmon\fP:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
cmd start <pid>
log start <pid>
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
A process is about to be stopped by \fBdmon\fP:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
cmd stop <pid>
log stop <pid>
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
A process has exited by its own means, or was terminated by the other means
different than \fBdmon\fP itself (e.g. by the kernel or the user):
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
cmd exit <pid> <status>
log exit <pid> <status>
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
The \fB<status>\fP field is numeric, and must be interpreted the same as the
\fIstatus\fP argument to the \fIwaitpid(2)\fP system call. Most of the time this is
the expected integer code passed to \fIexit(2)\fP, but this may not be true if
the process exits forcibly.
.sp
A signal is about to be sent to a process:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
cmd signal <pid> <signal>
log signal <pid> <signal>
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
The main monitored process timed out (when \fB\-t\fP is in effect):
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
cmd timeout <pid>
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
Process was paused or resumed due to system load constraints (when the
\fB\-l\fP and \fB\-L\fP options are in effect):
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
cmd pause <pid>
cmd resume <pid>
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.SH ENVIRONMENT
.sp
Additional options will be picked from the \fBDMON_OPTIONS\fP environment
variable, if defined. Any command line option can be specified this way.
Arguments read from the environment variable will be prepended to the ones
given in the command line, so they may still be overriden.
.sp
If the \fBDMON_LIST_MULTICALL_APPLETS\fP is defined and has a non\-zero value,
the list of applets compiled into a multicall binary will be printed out,
and the program will exit immediately.
.SH SEE ALSO
.sp
\fIdlog(8)\fP, \fIdslog(8)\fP, \fIrotlog(8)\fP, \fImultilog(8)\fP, \fIsupervise(8)\fP, \fIcron(8)\fP
.sp
\fI\%http://cr.yp.to/daemontools.html\fP
.SH AUTHOR
Adrian Perez <aperez@igalia.com>
.\" Generated by docutils manpage writer.
.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [v0.5.0] - 2020-08-27
### Added
- Support listing applets compiled into a multicall `dmon` binary when the
  `DMON_LIST_MULTICALL_APPLETS` environment variable is set and non-zero.

### Changed
- GNU Make is not required anymore, the included `Makefile` now works with
  the BSD variant as well, and probably others.

## [v0.4.5] - 2018-08-05
### Fixed
- Make it possible to build `dmon` with newer GCC versions.

## [v0.4.4] - 2016-10-30
### Fixed
- Correctly forward signals to the log process when using `-S`/`--log-sigs`.

## [v0.4.3] - 2016-10-13
### Added
- Allow setting the work directory with `-W`/`--work-dir` in the command line
  (or `work-dir` in a configuration file).

## [v0.4.2] - 2012-01-07
### Added
- Manual pages now also include the long command line options.
- `dlog` can now be instructed to prefix every log message with a given string,
  using the `-p`/`--prefix` command line option.
- `dlog`, `drlog`, and `dslog` have gained the `-i`/`--input-fd` command line
  option, which allows reading log messages from file descriptors different from
  the standard input.

### Fixed
- `dlog` and `drlog` now handle the `INT`, `TERM`, and `HUP` signals gracefully.

[Unreleased]: https://github.com/aperezdc/dmon/compare/v0.5.0...HEAD
[v0.5.0]: https://github.com/aperezdc/dmon/compare/v0.4.5...v0.5.0
[v0.4.5]: https://github.com/aperezdc/dmon/compare/v0.4.4...v0.4.5
[v0.4.4]: https://github.com/aperezdc/dmon/compare/v0.4.3...v0.4.4
[v0.4.3]: https://github.com/aperezdc/dmon/compare/v0.4.2...v0.4.3
[v0.4.2]: https://github.com/aperezdc/dmon/compare/v0.4.1...v0.4.2


dmon.c
/*
 * dmon.c
 * Copyright (C) 2010-2014 Adrian Perez <aperez@igalia.com>
 *
 * Distributed under terms of the MIT license.
 */

#define _GNU_SOURCE 1

#include "deps/cflag/cflag.h"
#include "deps/clog/clog.h"
#include "conf.h"
#include "task.h"
#include "util.h"
#include <assert.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <stdarg.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <fcntl.h>
#include <errno.h>
#include <time.h>


#if !(defined(MULTICALL) && MULTICALL)
# define dmon_main main
#endif /* !MULTICALL */


static int           log_fds[2]   = { -1, -1 };
static FILE         *status_file  = NULL;
static task_t        cmd_task     = TASK;
static task_t        log_task     = TASK;
static float         load_low     = 0.0f;
static float         load_high    = 0.0f;
static bool          success_exit = false;
static int           num_respawns = -1;
static bool          log_signals  = false;
static bool          cmd_signals  = false;
static unsigned long cmd_timeout  = 0;
static unsigned long cmd_interval = 0;
static int           check_child  = 0;
static int           running      = 1;
static int           paused       = 0;
static bool          nodaemon     = false;
static char         *status_path  = NULL;
static char         *pidfile_path = NULL;
static char         *workdir_path = NULL;


static const struct {
    const char *name;
    int         code;
} forward_signals[] = {
    { "CONT", SIGCONT },
    { "ALRM", SIGALRM },
    { "QUIT", SIGQUIT },
    { "USR1", SIGUSR1 },
    { "USR2", SIGUSR2 },
    { "HUP" , SIGHUP  },
    { "NONE", NO_SIGNAL },
    { "STOP", SIGSTOP },
    { "TERM", SIGTERM },
    { "INT" , SIGINT  },
    { "KILL", SIGKILL },
    { NULL  , NO_SIGNAL },
};


#define almost_zerof(_v)  ((_v) < 0.000000001f)

#define log_enabled   (log_fds[0] != -1)
#define load_enabled  (!almost_zerof (load_high))


__attribute__((format (printf, 1, 2)))
static inline void
_write_status (const char *fmt, ...)
{
    assert (status_file);

    va_list arg;
    va_start (arg, fmt);
    ssize_t r = vfprintf (status_file, fmt, arg);
    va_end (arg);

    if (r < 0)
        clog_warning("Writing to status file: %s.", ERRSTR);
}

#define write_status(...) \
    if (status_file != NULL) \
        _write_status(__VA_ARGS__)


#define task_action_dispatch_and_write_status(_what, _task)      \
    do {                                                          \
        int __pidafter__ = 0;                                     \
        if (status_file != NULL) {                                \
            switch ((_task)->action) {                            \
                case A_NONE:                                      \
                    break;                                        \
                case A_START:                                     \
                    _write_status ("%s start ", (_what));         \
                    __pidafter__ = 1;                             \
                    break;                                        \
                case A_STOP:                                      \
                    _write_status ("%s stop %li\n", (_what),      \
                                   (long) ((_task)->pid));        \
                    break;                                        \
                case A_SIGNAL:                                    \
                    _write_status ("%s signal %li %i\n", (_what), \
                                   (long) ((_task)->pid),         \
                                   (_task)->signal);              \
                    break;                                        \
            }                                                     \
        }                                                         \
        task_action_dispatch (_task);                             \
        if (__pidafter__ && status_file != NULL)                  \
            _write_status ("%li\n", (long) ((_task)->pid));       \
    } while (0)


const char*
signal_to_name (int signum)
{
    static const char *unknown = "(unknown)";
    int i = 0;

    while (forward_signals[i].name != NULL) {
        if (forward_signals[i].code == signum)
            return forward_signals[i].name;
        i++;
    }
    return unknown;
}

#if defined(__UCLIBC__)
#include <sys/sysinfo.h>
static int getloadavg(double *a, int n)
{
    struct sysinfo si;
    if (n <= 0) return n ? -1 : 0;
    if (n > 3) n = 3;

    if (sysinfo(&si) != 0) return -1;

    int i = 0;
    for (i=0; i<n; i++)
        a[i] = 1.0/(1<<SI_LOAD_SHIFT) * si.loads[i];

    return n;
}
#endif

static int
reap_and_check (void)
{
    clog_debug("Waiting for a children to reap...");

    int status;
    pid_t pid = waitpid (-1, &status, WNOHANG);

    if (pid == cmd_task.pid) {
        clog_debug("Reaped cmd process %d", pid);

        write_status ("cmd exit %li %i\n", (long) pid, status);

        cmd_task.pid = NO_PID;

        /*
         * If exit-on-success was request AND the process exited ok,
         * then we do not want to respawn, but to gracefully shutdown.
         */
        if (success_exit && WIFEXITED (status) && WEXITSTATUS (status) == 0) {
            clog_debug("cmd process ended successfully, will exit");
            running = 0;
        }
        else if (num_respawns == 0) {
            clog_debug("cmd process respawned max number of times, will exit");
            running = 0;
        } else {
            if (num_respawns > 0) {
                num_respawns -= 1;
            }
            task_action_queue (&cmd_task, A_START);
        }
        return status;
    }
    else if (log_enabled && pid == log_task.pid) {
        clog_debug("Reaped log process %i", pid);

        write_status ("log exit %li %i\n", (long) pid, status);

        log_task.pid = NO_PID;
        task_action_queue (&log_task, A_START);
    }
    else { 
        clog_debug("Reaped unknown process %i", pid);
    }

    /*
     * For cases where a return status is not meaningful (PIDs other than
     * that of the command being run) just return some invalid return code
     * value.
     */
    return -1;
}


static void
handle_signal (int signum)
{
    clog_debug("Got signal %i, %s", signum, signal_to_name(signum));

    /* Receiving INT/TERM signal will stop gracefully */
    if (signum == SIGINT || signum == SIGTERM) {
        running = 0;
        return;
    }

    /* Handle CHLD: check children */
    if (signum == SIGCHLD) {
        check_child = 1;
        return;
    }

    /*
     * If we have a maximum time to run the process, and we receive SIGALRM,
     * then the timeout was reached. As per signal(7) it is safe to kill(2)
     * the process from a signal handler, so we do that and then mark it for
     * respawning.
     */
    if (cmd_timeout && signum == SIGALRM) {
        write_status ("cmd timeout %li\n", (long) cmd_task.pid);
        task_action (&cmd_task, A_STOP);
        task_action_queue (&cmd_task, A_START);
        alarm (cmd_timeout);
        return;
    }

    unsigned i = 0;
    while (forward_signals[i].code != NO_SIGNAL) {
        if (signum == forward_signals[i++].code)
            break;
    }

    if (signum != NO_SIGNAL) {
        /* Try to forward signals */
        if (cmd_signals) {
            clog_debug("Delayed signal %i for cmd process", signum);
            task_action_queue (&cmd_task, A_SIGNAL);
            task_signal_queue (&cmd_task, signum);
        }
        if (log_signals && log_enabled) {
            clog_debug("Delayed signal %i for log process", signum);
            task_action_queue (&log_task, A_SIGNAL);
            task_signal_queue (&log_task, signum);
        }
    }
}



static void
setup_signals (void)
{
    unsigned i = 0;
    struct sigaction sa;

    sa.sa_handler = handle_signal;
    sa.sa_flags = SA_NOCLDSTOP;
    sigfillset(&sa.sa_mask);

    while (forward_signals[i].code!= NO_SIGNAL) {
        safe_sigaction (forward_signals[i].name,
                        forward_signals[i].code, &sa);
        i++;
    }

    safe_sigaction ("CHLD", SIGCHLD, &sa);
    safe_sigaction ("TERM", SIGTERM, &sa);
    safe_sigaction ("INT" , SIGINT , &sa);
}


static enum cflag_status
_environ_option(const struct cflag *spec, const char *arg)
{
    if (!spec)
        return CFLAG_NEEDS_ARG;

    const char *equalsign;
    if ((equalsign = strchr(arg, '=')) == NULL) {
        unsetenv(arg);
    } else {
        char *varname = strndup(arg, equalsign - arg);
        setenv(varname, equalsign + 1, 1);
        free(varname);
    }
    return CFLAG_OK;
}


static enum cflag_status
_rlimit_option(const struct cflag *spec, const char *arg)
{
    if (!spec)
        return CFLAG_NEEDS_ARG;

    long value;
    int limit;
    int status = parse_limit_arg(arg, &limit, &value);

    if (status < 0)
        return CFLAG_OK;
    if (status)
        return CFLAG_BAD_FORMAT;

    safe_setrlimit(limit, value);
    return CFLAG_OK;
}


static enum cflag_status
_store_uidgids_option (const struct cflag *spec, const char *arg)
{
    if (!spec)
        return CFLAG_NEEDS_ARG;

    char *arg_copy = strdup(arg);
    int status = parse_uidgids(arg_copy, spec->data);
    free(arg_copy);

    return status ? CFLAG_BAD_FORMAT : CFLAG_OK;
}

static enum cflag_status
_config_option(const struct cflag *spec, const char *arg)
{
    (void) arg;

    if (!spec)
        return CFLAG_NEEDS_ARG;

    assert(!"Unreachable");
    clog_error("Option --config/-C must be the first one specified");
    return CFLAG_BAD_FORMAT;
}

static const struct cflag dmon_options[] = {
    {
        .name = "config", .letter = 'C',
        .func = _config_option,
        .help =
            "Read options from the specified configuration file. If given, "
            "this option must be the first one in the command line.",
    },
    CFLAG(bool, "no-daemon", 'n', &nodaemon,
          "Do not daemonize, stay in foreground."),
    CFLAG(bool, "stderr-redir", 'e', &cmd_task.redir_errfd,
          "Redirect command's standard error stream to its standard "
          "output stream."),
    CFLAG(bool, "cmd-sigs", 's', &cmd_signals,
          "Forward signals to command process."),
    CFLAG(bool, "log-sigs", 'S', &log_signals,
          "Forward signals to log process."),
    CFLAG(bool, "once", '1', &success_exit,
          "Exit if command exits with a zero return code. The process "
          "will be still respawned when it exits with a non-zero code."),
    CFLAG(int, "max-respawns", 'm', &num_respawns,
          "Exit after max number of respawns no matter the exit code."),
    CFLAG(string, "write-info", 'I', &status_path,
          "Write information on process status to the given file. "
          "Sockets and FIFOs may be used."),
    CFLAG(string, "pid-file", 'p', &pidfile_path,
          "Write PID to a file in the given path."),
    CFLAG(string, "work-dir", 'W', &workdir_path,
          "Specify a working directory. All other specified relative paths "
          "have to be specified in relation with this directory."),
    CFLAG(float, "load-high", 'L', &load_high,
          "Stop process when system load surpasses the given value."),
    CFLAG(float, "load-low", 'l', &load_low,
          "Resume process execution when system load drops below the "
          "given value. If not given, defaults to half the value passed "
          "to '-L'."),
    CFLAG(timei, "timeout", 't', &cmd_timeout,
          "If command execution takes longer than the time specified "
          "the process will be killed and started again."),
    CFLAG(timei, "interval", 'i', &cmd_interval,
          "Time to wait between successful command executions. When "
          "exit code is non-zero, the interval is ignored and the "
          "command is executed again as soon as possible."),
    {
        .name = "environ", .letter = 'E',
        .func = _environ_option,
        .help =
            "Define an environment variable, or if no value is given, "
            "delete it. This option can be specified multiple times.",
    },
    {
        .name = "limit", .letter = 'r',
        .func = _rlimit_option,
        .help =
            "Sets a resource limit, given as 'name=value'. This option "
            "can be specified multiple times. Use '-r help' for a list.",
    },
    {
        .name = "cmd-user", .letter = 'u',
        .func = _store_uidgids_option,
        .data = &cmd_task.user,
        .help =
            "User and (optionally) groups to run the command as. Format "
            "is 'user[:group1[:group2[:...groupN]]]'.",
    },
    {
        .name = "log-user", .letter = 'U',
        .func = _store_uidgids_option,
        .data = &log_task.user,
        .help =
            "User and (optionally) groups to run the log process as. "
            "Format is 'user[:group1[:group2[:...groupN]]]'.",
    },
    CFLAG_HELP,
    CFLAG_END
};


int
dmon_main (int argc, char **argv)
{
    clog_init(NULL);

    FILE *pid_file = NULL;
    char *opts_env = NULL;

    /* Check for -C/--config given as first command line argument. */
    if (argc > 2 && ((argv[1][0] == '-' &&
                      argv[1][1] == 'C' &&
                      argv[1][2] == '\0') ||
                     !strcmp("--config", argv[1]))) {
        FILE *input = fopen(argv[2], "r");
        if (!input)
            die("%s: Cannot open file '%s', %s\n", argv[0], argv[2], ERRSTR);

        struct dbuf errmsg = DBUF_INIT;
        if (!conf_parse(input, dmon_options + 1, &errmsg))
            die("%s: Error parsing %s:%s\n", argv[0], argv[2], dbuf_str(&errmsg));

        replace_args_shift(2, &argc, &argv);
    }

    if ((opts_env = getenv ("DMON_OPTIONS")) != NULL)
        replace_args_string (opts_env, &argc, &argv);

    const char *argv0 = cflag_apply(dmon_options,
                                    "cmd [cmd-options] [ -- "
                                    "log-cmd [log-cmd-options]]",
                                    &argc, &argv);

    if (workdir_path) {
        if (chdir (workdir_path) != 0)
            die ("%s: Cannot use '%s' as work directory, %s\n", argv0, workdir_path, ERRSTR);
    }

    if (status_path) {
        int fd = open (status_path, O_WRONLY | O_CREAT | O_APPEND, 0666);
        if (fd < 0)
            die ("%s: Cannot open '%s' for writing, %s\n", argv0, status_path, ERRSTR);
        status_file = fdopen (fd, "w");
        setvbuf (status_file, NULL, _IOLBF, 0);
    }

    if (cmd_interval && success_exit)
        die ("%s: Options '-i' and '-1' cannot be used together.\n", argv0);

    if (load_enabled && almost_zerof (load_low))
        load_low = load_high / 2.0f;

    cmd_task.argv = argv;

    /* Skip over until "--" is found */
    unsigned i = 0;
    while (i < (unsigned) argc && strcmp (argv[i], "--") != 0) {
        cmd_task.argc++;
        i++;
    }

    /* There is a log command */
    if (i < (unsigned) argc && strcmp (argv[i], "--") == 0) {
        log_task.argc = argc - cmd_task.argc - 1;
        log_task.argv = argv + argc - log_task.argc;
        log_task.argv[log_task.argc] = NULL;
    }

    cmd_task.argv[cmd_task.argc] = NULL;

    if (log_task.argc > 0) {
        if (pipe (log_fds) != 0) {
            die ("%s: Cannot create pipe: %s\n", argv0, ERRSTR);
        }
        clog_debug("pipe_read = %i, pipe_write = %i\n", log_fds[0], log_fds[1]);
        fd_cloexec (log_fds[0]);
        fd_cloexec (log_fds[1]);
    }

    if (clog_debug_enabled) {
        char **xxargv = cmd_task.argv;
        fputs("cmd:", stderr);
        while (*xxargv) {
            fputc(' ', stderr);
            fputs(*xxargv++, stderr);
        }
        fputc('\n', stderr);
        if (log_enabled) {
            char **xxargv = log_task.argv;
            fputs("log:", stderr);
            while (*xxargv) {
                fputc(' ', stderr);
                fputs(*xxargv++, stderr);
            }
            fputc('\n', stderr);
        }
    }

    if (cmd_task.argc == 0)
        die ("%s: No command to run given.\n", argv0);

    if (pidfile_path) {
        int fd = open (pidfile_path, O_TRUNC | O_CREAT | O_WRONLY, 0666);
        if (fd < 0) {
            die ("%s: cannot open '%s' for writing, %s\n",
                 argv0, pidfile_path, ERRSTR);
        }
        pid_file = fdopen (fd, "w");
    }

    if (!nodaemon)
        become_daemon ();

    /* We have a valid file descriptor: write PID */
    if (pid_file) {
        if (fprintf (pid_file, "%li\n", (long) getpid ()) < 0)
            clog_warning("Writing to PID file: %s", ERRSTR);
        fclose (pid_file);
        pid_file = NULL;
    }

    setup_signals ();
    alarm (cmd_timeout);

    cmd_task.write_fd = log_fds[1];
    log_task.read_fd  = log_fds[0];

    int retcode = 0;
    while (running) {
        clog_debug(">>> loop iteration");
        if (check_child) {
            retcode = reap_and_check ();
            clog_debug("retcode = %d", retcode);

            /*
             * Wait the specified timeout but DO NOT use safe_sleep(): here
             * we want an interruptible sleep-wait so reaction to signals is
             * quick, which we definitely want for SIGINT/SIGTERM.
             */
            if (cmd_interval && !success_exit && retcode == 0 && num_respawns != 0) {
                int retval;
                struct timespec ts;
                ts.tv_sec = cmd_interval;
                ts.tv_nsec = 0;

                do {
                    retval = nanosleep (&ts, &ts);
                    clog_debug("nanosleep -> %i\n", retval);
                } while (retval == -1 && errno == EINTR && running);
            }

            /*
             * Either handling signals which interrupt the previous loop,
             * or reap_and_check() may request stopping on successful exit
             */
            if (!running) {
                task_action_queue (&cmd_task, A_NONE);
                break;
            }
        }

        task_action_dispatch_and_write_status ("cmd", &cmd_task);
        if (log_enabled)
            task_action_dispatch_and_write_status ("log", &log_task);

        if (load_enabled) {
            double load_cur;

            clog_debug("Checking load after sleeping 1s");
            interruptible_sleep (1);

            if (getloadavg (&load_cur, 1) == -1)
                clog_debug("getloadavg() failed: %s", ERRSTR);

            if (paused) {
                /* If the current load dropped below load_low -> resume */
                if (load_cur <= load_low) {
                    clog_debug("Resuming...");
                    task_signal (&cmd_task, SIGCONT);
                    write_status ("cmd resume %li\n", (long) cmd_task.pid);
                    paused = 0;
                }
            }
            else {
                /* If the load went above load_high -> pause */
                if (load_cur > load_high) {
                    clog_debug("Pausing...");
                    task_signal (&cmd_task, SIGSTOP);
                    write_status ("cmd pause %li\n", (long) cmd_task.pid);
                    paused = 1;
                }
            }
        }
        else {
            /* Wait for signals to arrive. */
            clog_debug("Waiting for signals to come...");
            pause ();
        }
    }

    clog_debug("Exiting gracefully...");

    if (cmd_task.pid != NO_PID) {
        write_status ("cmd stop %li\n", (long) cmd_task.pid);
        task_action (&cmd_task, A_STOP);
    }
    if (log_enabled && log_task.pid != NO_PID) {
        write_status ("log stop %li\n", (long) log_task.pid);
        task_action (&log_task, A_STOP);
    }

    if (status_file) {
        fclose (status_file);
        status_file = NULL;
    }

    if (WIFEXITED (retcode))
        exit (WEXITSTATUS (retcode));

    exit (EXIT_FAILURE);
}

/* vim: expandtab shiftwidth=4 tabstop=4
 */
